### PR TITLE
[Merged by Bors] - feat: port Algebra.Ring.Defs

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
Tracking mathlib commit: [39af7d3bf61a98e928812dbc3e16f4ea8b795ca3](https://github.com/leanprover-community/mathlib/commit/39af7d3bf61a98e928812dbc3e16f4ea8b795ca3)

### CODE REVIEW

**Comment by semorrison** [comment]
See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/mathlib4.23655 for the discussion about the failure of `ring` at this point.

**Comment by semorrison** [comment]
Could you please fix the long lines?

File: Mathlib/Algebra/Ring/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+/-- A unital but not-necessarily-associative semiring. -/
+class NonAssocSemiring (α : Type u) extends NonUnitalNonAssocSemiring α, MulZeroOneClass α,
+    AddCommMonoidWithOne α
+#align non_assoc_semiring NonAssocSemiring
+
+class Semiring (α : Type u) extends NonUnitalSemiring α, NonAssocSemiring α, MonoidWithZero α
+#align semiring Semiring
+
+-- Porting note: no OfNat α 2
+/-
```
Comment (semorrison) [inline Mathlib/Algebra/Ring/Defs.lean:112]: Please leave things like this uncommented. For things that still need to be ported, error messages in the build are more helpful than "Porting note", which should be intended for notes that will be committed as part of the PR.

File: Mathlib/Algebra/Ring/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+    CommMonoidWithZero α :=
+  { inferInstanceAs (CommMonoid α), inferInstanceAs (CommSemiring α) with }
+#align comm_semiring.to_comm_monoid_with_zero CommSemiring.toCommMonoidWithZero
+
+section CommSemiring
+
+variable [CommSemiring α] {a b c : α}
+
+set_option warningAsError false
+
```
Comment (semorrison) [inline Mathlib/Algebra/Ring/Defs.lean:239]: No need for this: the sorry needs to be fixed before we merge!

**Comment by Ruben-VandeVelde** [comment]
Thanks for the help, should be ready for review now

File: Mathlib/Algebra/Ring/Defs.lean [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+#align non_assoc_semiring NonAssocSemiring
+
+class Semiring (α : Type u) extends NonUnitalSemiring α, NonAssocSemiring α, MonoidWithZero α
+#align semiring Semiring
+
+section AddMonoidWithOne
+
+variable [AddMonoidWithOne α]
+
+-- TODO: Move to `Data.Nat.Cast.Defs`?
```
Comment (semorrison) [inline Mathlib/Algebra/Ring/Defs.lean:119]: :+1:

File: Mathlib/Algebra/Ring/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+
+variable [Semiring α]
+
+-- Porting note: was [has_add α] [mul_one_class α] [right_distrib_class α]
+theorem two_mul (n : α) : 2 * n = n + n :=
+  (congrArg₂ _ one_add_one_eq_two.symm rfl).trans <| (right_distrib 1 1 n).trans (by rw [one_mul])
+#align two_mul two_mul
+
+section deprecated
+set_option linter.deprecated false
```
Comment (semorrison) [inline Mathlib/Algebra/Ring/Defs.lean:160]: If it's only for one command, you can use

`set_option linter.deprecated false in ...`

**Comment by semorrison** [comment]
Looks good otherwise.

bors d+

**Comment by Ruben-VandeVelde** [comment]
Thanks!

bors r+

File: Mathlib/Algebra/Ring/Defs.lean [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+/-- An associative but not-necessarily unital ring. -/
+class NonUnitalRing (α : Type _) extends NonUnitalNonAssocRing α, NonUnitalSemiring α
+#align non_unital_ring NonUnitalRing
+
+/-- A unital but not-necessarily-associative ring. -/
+class NonAssocRing (α : Type _) extends NonUnitalNonAssocRing α, NonAssocSemiring α,
+    AddGroupWithOne α
+#align non_assoc_ring NonAssocRing
+
+class Ring (R : Type u) extends Semiring R, AddCommGroup R, AddGroupWithOne R
```
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Defs.lean:326]: This doesn't match the version in Lean 3...

[Zulip Thread](
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/ring.20vs.20Ring/near/322876462)

**Pull request closed without merging**